beacon-chain/sync/service.go

```diff
@@ -0,0 +1,134 @@
+//     1. Receive a block hash from a peer
+//     2. Request the block for the hash from the network
+//     3. Receive the block
+//     4. Send the block to the local chain to validate
+//
+//  In addition, Service will handle the following responsibilities:
+//     *  Decide which messages are forwarded to other peers
+//     *  Filter redundant data and unwanted data
+//     *  Drop peers that send invalid data
+//     *  Trottle incoming requests
+type Service struct {
+	ctx            context.Context
+	cancel         context.CancelFunc
+	networkService NetworkService
+	chainService   ChainService
+	hashBuf        chan hash.Hash
+	blockBuf       chan *types.Block
+}
+
+// Config allows the channel's buffer sizes to be changed
+type Config struct {
+	HashBufferSize int
+	BlockBufferSize int
+}
+
+// DefaultConfig provides the default configuration for a sync service
+func DefaultConfig() Config {
+	return Config { 100, 100 }
+}
+
+// NetworkService is the interface for the p2p network.
+type NetworkService interface {
+	BroadcastBlockHash(hash.Hash) error
+	BroadcastBlock(*types.Block) error
+	RequestBlock(hash.Hash) error
+}
+
+// ChainService is the interface for the local beacon chain.
+type ChainService interface {
+	ProcessBlock(*types.Block) error
+	ContainsBlock(hash.Hash) bool
+}
+
+// NewSyncService accepts a context and returns a new Service.
+func NewSyncService(ctx context.Context, cfg Config) *Service {
+	ctx, cancel := context.WithCancel(ctx)
+	return &Service{
+		ctx:      ctx,
+		cancel:   cancel,
+		hashBuf:  make(chan hash.Hash, cfg.HashBufferSize),
+		blockBuf: make(chan *types.Block, cfg.BlockBufferSize),
+	}
+}
+
+// SetNetworkService sets a concrete value for the p2p layer.
+func (ss *Service) SetNetworkService(ps NetworkService) {
+	ss.networkService = ps
+}
+
+// SetChainService sets a concrete value for the local beacon chain.
+func (ss *Service) SetChainService(cs ChainService) {
+	ss.chainService = cs
+}
+
+// Start begins the block processing goroutine.
+func (ss *Service) Start() {
+	log.Info("Starting service")
+	go run(ss.ctx.Done(), ss.hashBuf, ss.blockBuf, ss.networkService, ss.chainService)
+}
+
+// Stop kills the block processing goroutine, but does not wait until the goroutine exits.
+func (ss *Service) Stop() error {
+	log.Info("Stopping service")
+	ss.cancel()
+	return nil
+}
+
+// ProcessBlockHash accepts a block hash.
+// New hashes are forwarded to other peers in the network (unimplemented), and
+// the contents of the block are requested if the local chain doesn't have the block.
+func (ss *Service) ProcessBlockHash(h hash.Hash) {
+	if ss.chainService.ContainsBlock(h) {
+		return
+	}
+
+	ss.hashBuf <- h
+	ss.networkService.BroadcastBlockHash(h)
+}
+
+// ProcessBlock accepts a block to potentially be included in the local chain.
+// The service will filter blocks that have not been requested (unimplemented).
+func (ss *Service) ProcessBlock(b *types.Block) error {
```

Comment:
I think this should be renamed to something like `ReceiveIncomingBlock`, given that `ProcessBlock` is something entirely different and it's part of the `blockchain` package. This function in sync should receive them via p2p and send them to a feed the chainService subscribes to. The chainService will then run a processBlocks goroutine which will check for validity conditions, calculate active state, and store in a trie.